rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,367 @@
+
+For GCP, we will map the [following JWT claims](https://cloud.google.com/compute/docs/instances/verifying-instance-identity#token_format) to configuration values:
+
+- `sub`: subject
+- `google.compute_engine.project_id`: project_id
+- `google.compute_engine.project_number`: project_number
+- `google.compute_engine.instance_name`: instance_name
+- `google.compute_engine.instance_id`: instance_id
+- `google.compute_engine.zone`: zone
+
+To ensure that we guide users to creating secure configurations, we will also ensure that at least one of the following fields is included in each allow block:
+
+- project_id
+- project_number
+- subject
+
+This ensures that a token produced in another GCP project cannot be used against their Teleport cluster.
+
+#### Extracting claims as metadata for generated credentials
+
+The JWTs issued by providers often contain claims that would be useful when auditing actions. We can extract these claims, and embed them into the certificates during registration. This additional metadata can then be audit logged when the certificates are used, allowing actions to be attributed to specific CI runs or individual VMs.
+
+In order to implement OIDC joining in a timely manner, we should consider this out of scope for this initial implementation.
+
+### Node support
+
+Node here not only refers to a Teleport node, but also to various other participants within a Teleport cluster (e.g tbot, kube agent etc).
+
+We will need to support collecting the token from the environment. This will differ on each platform. Some offer the token via a metadata service, and others directly inject it via an environment variable. Where possible, we should encourage the user to configure the token to be generated with an audience of their Teleport cluster, however, not all providers support this (e.g GitLab CI/CD).
+
+For GCP, a HTTP request is made to a metadata service. In this request, a query parameter controls the audience of the generated token. E.g
+
+```
+GET http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity?audience=https://noah.teleport.sh&format=full
+```
+
+### Security Considerations
+
+#### Secure authorization rules
+
+It is imperative that we validate the user's provided `allow` rules so that they do not accidentally allow tokens produced in other projects to be used against their teleport cluster. For example, if they provided an `allow` rule that only included `instance_name`, a malicious actor would be able to create an instance with the same name, and use this to generate a valid token.
+
+#### MITM of the connection to the issuer
+
+In order to verify the JWT, we have to fetch the public key from the issuers's JWKS endpoint. If this connection is not secured (e.g HTTPS), it would be possible for a malicious actor to intercept the request and return a public key they've used to sign the JWT with.
+
+We should require that the configured issuer URL is HTTPS to mitigate this.
```

Comment:
This is certainly an interesting option. I'm not aware that Google or Github are sending HPKP headers for their well-known OIDC endpoint, but we could allow the user to manually configure the certificate they expect Google or Github to be using. I'm a little worried that this could be extremely fragile as I'm unsure of how often OP's are rotating the certificate used for the TLS on these endpoints or if they use a variety of certificates at any one time.

Reply:
It won't be practical for GCP (IIRC there's a daily rotation) but it could be an option for the future

Reply:
Sure ! It's certainly a very interesting option if we ever allow custom providers where users may well know exactly which certificate they are expecting to see! I'll add it as a future consideration to make sure we have it noted down.